File: src/imanage/btime_utils.py
"""
btime_utils — macOS btime (st_birthtime) 保全ユーティリティ

ポリシー: imanage はいかなる操作においてもファイルの btime を変更してはならない。
"""
import os, ctypes, ctypes.util, struct, shutil
from contextlib import contextmanager

ATTR_BIT_MAP_COUNT = 5
ATTR_CMN_CRTIME    = 0x00000200


class _attrlist(ctypes.Structure):
    _fields_ = [
        ("bitmapcount", ctypes.c_uint16), ("reserved", ctypes.c_uint16),
        ("commonattr",  ctypes.c_uint32), ("volattr",  ctypes.c_uint32),
        ("dirattr",     ctypes.c_uint32), ("fileattr", ctypes.c_uint32),
        ("forkattr",    ctypes.c_uint32),
    ]


def get_btime(path: str) -> float | None:
    try:
        return os.stat(path).st_birthtime
    except AttributeError:
        return None


def set_btime(path: str, btime: float) -> None:
    """macOS 専用: setattrlist(2) で birthtime を設定する"""
    libc = ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True)
    attrs = _attrlist()
    attrs.bitmapcount = ATTR_BIT_MAP_COUNT
    attrs.commonattr  = ATTR_CMN_CRTIME
    sec, nsec = int(btime), int((btime - int(btime)) * 1_000_000_000)
    buf = struct.pack("qq", sec, nsec)  # timespec: tv_sec + tv_nsec (各 8 bytes)
    ret = libc.setattrlist(path.encode("utf-8"), ctypes.byref(attrs), buf, len(buf), 0)
    if ret != 0:
        errno = ctypes.get_errno()
        raise OSError(errno, os.strerror(errno), path)


@contextmanager
def preserve_btime(path: str):
    """btime を保存してブロック終了後に復元するコンテキストマネージャ"""
    btime = get_btime(path)
    try:
        yield
    finally:
        if btime is not None:
            try:
                set_btime(path, btime)
            except Exception:
                pass  # 復元失敗は silent fail（処理継続を優先）


def btime_safe_move(src: str, dst: str) -> None:
    """btime を保持したまま shutil.move を実行する"""
    btime = get_btime(src)
    shutil.move(src, dst)
    if btime is not None:
        final = os.path.join(dst, os.path.basename(src)) if os.path.isdir(dst) else dst
        try:
            set_btime(final, btime)
        except Exception:
            pass

File: app/Imanage/Core/BTimeUtils.swift
import Foundation
import Darwin

// MARK: - BTimeUtils
// btime_utils.py の移植。
// ポリシー: imanage はいかなる操作においてもファイルの btime を変更してはならない。
// すべてのファイル移動はこのモジュール経由で行い、移動後に btime を復元する。

enum BTimeUtils {

    // MARK: - btime 取得

    /// stat(2) 経由で st_birthtime を timespec として返す。
    /// btime が取得できない場合は nil。
    static func getBirthTime(atPath path: String) -> timespec? {
        var st = stat()
        guard stat(path, &st) == 0 else { return nil }
        return st.st_birthtimespec
    }

    // MARK: - btime 設定

    /// setattrlist(2) で birthtime を設定する（macOS 専用）。
    /// 失敗時は OSError をスロー。
    static func setBirthTime(atPath path: String, to time: timespec) throws {
        var attrs = attrlist()
        attrs.bitmapcount = u_short(ATTR_BIT_MAP_COUNT)
        attrs.commonattr  = attrgroup_t(ATTR_CMN_CRTIME)

        // timespec (tv_sec: Int64 + tv_nsec: Int64) を連続メモリとして渡す
        var ts = time
        let ret = withUnsafeMutableBytes(of: &ts) { buf -> Int32 in
            setattrlist(path, &attrs, buf.baseAddress!, buf.count, 0)
        }
        if ret != 0 {
            throw NSError(
                domain: NSPOSIXErrorDomain,
                code: Int(errno),
                userInfo: [NSLocalizedDescriptionKey: String(cString: strerror(errno))]
            )
        }
    }

    // MARK: - btime 保持ファイル移動

    /// btime を保持したまま from を to へ移動する。
    /// to はフルパス（ファイル名込み）。親ディレクトリは存在する前提。
    /// to に同名ファイルが存在する場合は何もせず false を返す。
    /// 失敗時は throw する。
    @discardableResult
    static func safeMove(from src: URL, to dest: URL) throws -> Bool {
        // 移動先に同名ファイルが存在する場合はスキップ
        guard !FileManager.default.fileExists(atPath: dest.path) else {
            return false
        }

        let btime = getBirthTime(atPath: src.path)

        try FileManager.default.moveItem(at: src, to: dest)

        // 移動後に btime を復元（失敗は silent）
        if let bt = btime {
            try? setBirthTime(atPath: dest.path, to: bt)
        }
        return true
    }
}
